### PR TITLE
RFC: Implement array textures

### PR DESCRIPTION
**Pull request by linkmauve**
I couldn’t find any good way to pass the number of layers, hence the RFC.

It’s also missing 1D and cubemap array textures, but those can come later.

Otherwise it works.

### CODE REVIEW

**Comment by phaazon** [comment]
Thanks, I’ll have a look tonight.

**Comment by linkmauve** [comment]
Actually I think I should use `Dim3` for 2D array textures, and `Dim2` for 1D array textures, this would solve the main API issue.

**Comment by linkmauve** [comment]
This now implements 1D and 2D array textures as Dim2 and Dim3.

Cubemaps are left as an exercise to the reader.

**Comment by phaazon** [comment]
Okay, I’ll review that.

File: luminance/src/pipeline.rs [hunk 1]
```diff
@@ -515,30 +515,46 @@ where
...
+      (Layering::Layered, PxType::Unsigned, Dim::Dim2) => Type::UISampler1DArray,
+      (Layering::Layered, PxType::Floating, Dim::Dim2) => Type::Sampler1DArray,
+
+      (Layering::Layered, PxType::NormIntegral, Dim::Dim3) => Type::Sampler2DArray,
+      (Layering::Layered, PxType::NormUnsigned, Dim::Dim3) => Type::Sampler2DArray,
+      (Layering::Layered, PxType::Integral, Dim::Dim3) => Type::ISampler2DArray,
+      (Layering::Layered, PxType::Unsigned, Dim::Dim3) => Type::UISampler2DArray,
+      (Layering::Layered, PxType::Floating, Dim::Dim3) => Type::Sampler2DArray,
+
+      (Layering::Layered, _, Dim::Cubemap) => todo!("cubemap array samplers"),
```
Comment (phaazon) [inline luminance/src/pipeline.rs]: I didn’t know about that macro! Pretty cool!

File: luminance/src/texture.rs [hunk 2]
```diff
@@ -781,9 +781,9 @@ pub(crate) fn opengl_target(l: Layering, d: Dim) -> GLenum {
       Dim::Cubemap => gl::TEXTURE_CUBE_MAP,
     },
     Layering::Layered => match d {
-      Dim::Dim1 => gl::TEXTURE_1D_ARRAY,
-      Dim::Dim2 => gl::TEXTURE_2D_ARRAY,
-      Dim::Dim3 => unimplemented!("3D textures array not supported"),
+      Dim::Dim1 => unimplemented!("0D textures array not supported"),
```
Comment (phaazon) [inline luminance/src/texture.rs]: I… don’t get that change. Layered 1D textures are not 0D. They are 1D textures organized in layers.

File: luminance/src/texture.rs [hunk 3]
```diff
@@ -781,9 +781,9 @@ pub(crate) fn opengl_target(l: Layering, d: Dim) -> GLenum {
...
+      Dim::Dim2 => gl::TEXTURE_1D_ARRAY,
```
Comment (phaazon) [inline luminance/src/texture.rs]: Same thing here and below.

File: luminance/src/texture.rs [hunk 4]
```diff
@@ -858,6 +860,35 @@ where
           Ok(())
         }
 
+        // 1D array texture
```
Comment (phaazon) [inline luminance/src/texture.rs]: Eww nah. 1D array texture is a `Layered` `Dim1`.

File: luminance/src/texture.rs [hunk 5]
```diff
@@ -1203,7 +1236,41 @@ where
         },
       },
 
-      Layering::Layered => unimplemented!("Layering::Layered not implemented yet"),
+      Layering::Layered => match D::dim() {
+        Dim::Dim1 => panic!("Array textures can’t have 0 dimensions"),
```
Comment (phaazon) [inline luminance/src/texture.rs]: Yes, I agree. Hence, why have you decided to map `1 -> 0`?

**Comment by linkmauve** [comment]
Alright so you disagree with https://github.com/phaazon/luminance-rs/pull/291#issuecomment-575908653

If so, please tell me how to index the layer from a 2D array texture, which already has two dimensions, and a third one must then exist in order to index the layer.

**Comment by phaazon** [comment]
So, either we add a new dimensionality (and we get rid of `Layerable`?), so that we can provide a type for the offset / size (as in [here](https://github.com/phaazon/luminance-rs/blob/master/luminance/src/texture.rs#L240), or we allow specific functions to take an extra parameter on layered textures but not flat ones.

I think I’d rather like the first one. Especially because there is a bijection between the dimensionality of a texture (i.e. `Dim1`) and its layering (right now it’s always `Flat`). So if add `Dim1Array`, we don’t lose any information and we can get rid of `Layerable`, and a type variable.

**Comment by linkmauve** [comment]
So basically encode `Layerable` into `Dim`?  How do you plan on supporting cubemap arrays?

**Comment by phaazon** [comment]
We just add a `CubemapArray` variant.

**Comment by linkmauve** [comment]
Here, I got rid of `Layering::Layerable`, now every array texture is `Layering::Flat` with a new `Dim::Dim1Array` and `Dim::Dim2Array`!

This should be good to go. :)

**Comment by phaazon** [comment]
Great, thanks for your work! I’ll review and hope we can merge ASAP! :)

File: luminance/src/shader/program.rs [hunk 6]
```diff
@@ -810,12 +822,18 @@ impl fmt::Display for Type {
       Type::ISampler1D => f.write_str("isampler1D"),
       Type::ISampler2D => f.write_str("isampler2D"),
       Type::ISampler3D => f.write_str("isampler3D"),
+      Type::ISampler1DArray => f.write_str("isampler1DArray"),
+      Type::ISampler2DArray => f.write_str("isampler2DArray"),
       Type::UISampler1D => f.write_str("uSampler1D"),
       Type::UISampler2D => f.write_str("uSampler2D"),
       Type::UISampler3D => f.write_str("uSampler3D"),
+      Type::UISampler1DArray => f.write_str("uSampler1DArray"),
```
Comment (phaazon) [inline luminance/src/shader/program.rs]: May you please do as the way it’s done everywhere else? `usampler1DArray`.

File: luminance/src/shader/program.rs [hunk 7]
```diff
@@ -810,12 +822,18 @@ impl fmt::Display for Type {
...
+      Type::UISampler2DArray => f.write_str("uSampler2DArray"),
```
Comment (phaazon) [inline luminance/src/shader/program.rs]: Ditto.

File: luminance/src/shader/stage.rs [hunk 8]
```diff
@@ -87,7 +87,7 @@ impl Stage {
 
         gl::DeleteShader(handle);
 
-        log.set_len(log_len as usize);
+        log.set_len(log_len as usize - 1);
```
Comment (phaazon) [inline luminance/src/shader/stage.rs]: I really am not sure about that.

Screenshot 2020-02-06 at 15 43 29

File: luminance/src/shader/program.rs (hunk 6 above)
Comment (linkmauve) [reply luminance/src/shader/program.rs]: Just one line above it’s `uSampler3D`, I just copied how it is done.  Good catch, those were invalid. :)

File: luminance/src/shader/stage.rs (hunk 8 above)
Comment (linkmauve) [reply luminance/src/shader/stage.rs]: Indeed, I’ll have a look at why I got a null terminator then, probably a Mesa bug. Thanks!

**Comment by linkmauve** [comment]
This is now rebased and ready to go!

**Pull request merged**